### DISCUSSION TITLE
Allow changing of a server's egg after creation

### DISCUSSION THREAD

**Discussion by notAreYouScared** [pelican-dev/panel#430] [Ideas]
Allow changing a server's egg after the server has been created...

This should nuke the startup command and replace it with the new egg startup command

This should nuke all the egg variables assigned to the server and replace them with the new egg variables.

**Comment by rmartinoscar** [comment]
Could this keep current egg variables in common and without default values like pterodactyl did ?
Would be nice for Minecraft eggs for example if you want to switch from Spigot to Paper but keep same VERSION instead of rewritting it

**Comment by ItsNeil17** [comment]
Has anyone taken this up? I can do it.

After the egg gets changed, I believe it should re-run the install scripts?

**Comment by lancepioch** [comment]
Honestly, I can't see this being used too much or very useful. The examples I've seen and been told seem to be edge cases. If the eggs don't line up with their variables, then it's already a bit of hassle.

I don't see how you just stop the current server, create a new one with the different egg, and use that. If you need to upgrade the egg, that's already built in.

However my stance isn't very strong on this one. Message me or hop on Discord and tell me how you'll use this! For now I'll just convert this to a discussion.

**Reply by notAreYouScared** [reply]
I can't entirely agree with this. We shouldn't be plucking features from pterodactyl. 

With your logic of. 
> I don't see how you just stop the current server, create a new one with the different egg, and use that. If you need to upgrade the egg, that's already built in.

Why should we re-implement server transfers? You could use the same logic and tell the end user to stop the server, make a backup, download said backup, create a server on the new node, and then upload and extract the backup on the newly created server and redo the variables.

> If the eggs don't line up with their variables, then it's already a bit of hassle.

Pterodactyl nukes all the variables... We shouldn't expect them to line up... remove the current variables and replace them with the newly selected egg's variables.

**Comment by TekExplorer** [comment]
We could add a new configurator that lets you create from-to definitions that tell the panel how to transition between eggs.

If a from-to is defined in both directions, you can swap between those eggs, if enabled.

These definitions need to handle all variables that don't exist on the new egg, and new variables that don't exist on the old egg.

Any number of actions could occur - sharing "equivalent" variable values, or requiring the user to specify a value that didn't exist before you can switch to it, and "orphaned" values remain stored for when the user transitions back.

Admins either need to specify *all* required variables on all eggs, or only specify some required values on one or some eggs, and the user will be required to set them to use them.

Thoughts? This could be pretty powerful.

**Reply by notAreYouScared** [reply]
This isn't for the end user to change, this is for the admin side only.

**Reply by TekExplorer** [reply]
I don't see how that changes my proposal 🤷

Plus, it makes it so that egg changers between ex paper, forge, fabric, etc utterly trivial.

**Comment by rmartinoscar** [comment]
This would be nice for product base sevice provisioning like WHMCS yes you can use Custom Fields to allow between a range of Eggs but most of the time customers will just come at you asking for a change.

Also when you change an egg in my opinion you wanna keep your database(s), your credentials and your allocation(s).

**Comment by HerrSammyDE** [comment]
It would be a great advantage if you didn't have to create a new server to be able to play a different game
Please add again! 🚀

**Comment by rmartinoscar** [comment]
Closed with https://github.com/pelican-dev/panel/pull/811
